app/controllers/franklin_alma_controller.rb

```diff
@@ -104,52 +104,54 @@ def holding_details
+  # Also returns received coverage value
+  # @todo wrap API calls
+  # Params can include:
+  #  - portfolio_pid
+  #  - collection_id
+  #  - coverage
+  #  - public_note
+  # Data is queried for in this order:
+  #  1. Portfolio API
+  #  2. Service API
+  #  3. Collection API
   def portfolio_details
-    portfolio_pid = params['portfolio_pid']
-    collection_id = params['collection_id']
-    api_key_param = "apikey=#{ENV['ALMA_API_KEY']}"
-    url_params = { collection_id: collection_id }
-    coverage = params['coverage']
-
-    # we also get this from availability API. Opportunity for improvement?
-    public_note = nil
-    authentication_note = nil
-
-    collection_url = "https://api-na.hosted.exlibrisgroup.com/almaws/v1/electronic/e-collections/%{collection_id}?#{api_key_param}"
-    services_url = "https://api-na.hosted.exlibrisgroup.com/almaws/v1/electronic/e-collections/%{collection_id}/e-services?#{api_key_param}"
-    portfolio_url = "https://api-na.hosted.exlibrisgroup.com/almaws/v1/electronic/e-collections/%{collection_id}/e-services/%{service_id}/portfolios/%{portfolio_id}?#{api_key_param}"
-
-    api_response = HTTParty.get(
-      services_url % url_params,
-      headers: { 'Accept' => 'application/json' }
-    )
-
-    # 'electronic_service' is not set for e-portfolios et al
-    api_response['electronic_service'].each do |e|
-      portfolio_response = HTTParty.get(portfolio_url % url_params.merge(:portfolio_id => portfolio_pid, :service_id => e['id']), :headers => {'Accept' => 'application/json'})
-      public_note ||= portfolio_response['public_note'].presence
-      authentication_note ||= portfolio_response['authentication_note'].presence
-
-      if public_note.nil? || authentication_note.nil?
-        service_response = HTTParty.get(e['link'] + "?#{api_key_param}", :headers => {'Accept' => 'application/json'})
-        public_note ||= service_response['public_note'].presence
-        authentication_note ||= service_response['authentication_note'].presence
-      end
+    alma_api_base_path = 'https://api-na.hosted.exlibrisgroup.com/almaws/v1'
+    request_headers = { 'Accept' => 'application/json' }
+
+    # TODO: params can have a public_note value, not to mention coverage. where are those values coming from?
+    # the Availability API call made earlier? "we also get this from availability API. Opportunity for improvement?"
+
+    # Get the right Service ID from the Bibs API
+    bib_portfolio_url = "#{alma_api_base_path}/bibs/#{params[:mms_id]}/portfolios/#{params[:portfolio_pid]}?#{api_key_param}"
+    bib_portfolio_response = HTTParty.get(bib_portfolio_url, headers: request_headers)
+    service_id = bib_portfolio_response.dig 'electronic_collection', 'service', 'value'
+
+    # The portfolio is the most trustworthy source for details
+    portfolio_url = "#{alma_api_base_path}/electronic/e-collections/#{params[:collection_id]}/e-services/#{service_id}/portfolios/#{params[:portfolio_pid]}?#{api_key_param}"
```

Comment:
Should there be a check here in case `params[:collection_id]` is nil?

Reply:
yes, will do